nullaway/src/main/java/com/uber/nullaway/generics/ConstraintSolverImpl.java

```diff
@@ -0,0 +1,303 @@
+          lhsTypeArg.accept(this, rhsTypeArg);
+          rhsTypeArg.accept(this, lhsTypeArg);
+        }
+      }
+      // if supertype is not a ClassType, we still call visitType to handle the case where
+      // supertype is a TypeVar
+      return visitType(subtype, supertype);
+    }
+
+    @Override
+    public Void visitArrayType(Type.ArrayType subtype, Type supertype) {
+      if (supertype instanceof Type.ArrayType) {
+        Type.ArrayType superArrayType = (Type.ArrayType) supertype;
+        // recursing, so set localVariableType to false
+        localVariableType = false;
+        Type subtypeComponentType = subtype.elemtype;
+        Type superComponentType = superArrayType.elemtype;
+        // constrain in both directions
+        subtypeComponentType.accept(this, superComponentType);
+        superComponentType.accept(this, subtypeComponentType);
+      }
+      // if supertype is not an ArrayType, we still call visitType to handle the case where
+      // supertype is a TypeVar
+      return visitType(subtype, supertype);
+    }
+
+    @Override
+    public Void visitTypeVar(TypeVar subtype, Type supertype) {
+      if (!localVariableType) {
+        directlyConstrainTypePair(subtype, supertype);
+      }
+      return visitType(subtype, supertype);
+    }
+  }
+
+  @Override
+  public Map<TypeVariable, InferredNullability> solve() throws UnsatisfiableConstraintsException {
+    /* ---------- work-list propagation of nullability ---------- */
+    Deque<TypeVariable> work = new ArrayDeque<>();
+    vars.forEach(
+        (tv, st) -> {
+          if (st.nullness != NullnessState.UNKNOWN) {
+            work.add(tv);
+          }
+        });
+
+    while (!work.isEmpty()) {
+      TypeVariable tv = work.removeFirst();
+      VarState st = castToNonNull(vars.get(tv));
+
+      switch (st.nullness) {
+        case NONNULL:
+          /* S <: tv  &  tv NONNULL  ⇒  S NONNULL */
+          for (TypeVariable sub : st.subtypes) {
+            if (updateNullness(sub, NullnessState.NONNULL)) {
+              work.add(sub);
+            }
+          }
+          break;
+
+        case NULLABLE:
+          /* tv <: T  &  tv NULLABLE  ⇒  T NULLABLE */
+          for (TypeVariable sup : st.supertypes) {
+            if (updateNullness(sup, NullnessState.NULLABLE)) {
+              work.add(sup);
+            }
+          }
+          break;
+
+        default: // UNKNOWN
+          throw new RuntimeException("Unexpected nullness state: " + st.nullness + " for " + tv);
+      }
+    }
+
+    /* ---------- build final solution map ---------- */
+    Map<TypeVariable, InferredNullability> result = new HashMap<>();
+    vars.forEach(
+        (tv, st) -> {
+          // if the nullness state is UNKNOWN, set it to NONNULL arbitrarily
+          // TODO does this matter?  should we use NULLABLE instead?
+          NullnessState n =
+              (st.nullness == NullnessState.UNKNOWN) ? NullnessState.NONNULL : st.nullness;
```

Comment:
Isn't this redundant with the check below that will return `InferredNullability.NONNULL` for anything but `NullnessState.NULLABLE`?

Reply:
Yes, you're right!  Fixed in c9d7e081